File: Problems/Problem_1195_Fizz_Buzz_Multithreaded/java/FizzBuzz.java
package Problems.Problem_1195_Fizz_Buzz_Multithreaded.java;

import java.util.function.IntConsumer;
import java.util.stream.IntStream;

class FizzBuzz {
    private int n;
    private int current = 1;

    public FizzBuzz(int n) {
        this.n = n;
    }

    // printFizz.run() outputs "fizz".
    public synchronized void fizz(Runnable printFizz) throws InterruptedException {
        while (current <= n) {
            if (current % 3 == 0 && current % 5 != 0) {
                printFizz.run();
                current++;
                notifyAll();
            } else {
                wait();
            }
        }
    }

    // printBuzz.run() outputs "buzz".
    public synchronized void buzz(Runnable printBuzz) throws InterruptedException {
        while (current <= n) {
            if (current % 5 == 0 && current % 3 != 0) {
                printBuzz.run();
                current++;
                notifyAll();
            } else {
                wait();
            }
        }
    }

    // printFizzBuzz.run() outputs "fizzbuzz".
    public synchronized void fizzbuzz(Runnable printFizzBuzz) throws InterruptedException {
        while (current <= n) {
            if (current % 3 == 0 && current % 5 == 0) {
                printFizzBuzz.run();
                current++;
                notifyAll();
            } else {
                wait();
            }
        }
    }

    // printNumber.accept(x) outputs "x", where x is an integer.
    public synchronized void number(IntConsumer printNumber) throws InterruptedException {
        while (current <= n) {
            if (current % 3 != 0 && current % 5 != 0) {
                printNumber.accept(current);
                current++;
                notifyAll();
            } else {
                wait();
            }
        }
    }

    public static void main(String[] args) {
        Runnable printFizz = new Runnable() {
            public void run() {
                System.out.println("fizz");
            }
        };
        Runnable printBuzz = new Runnable() {
            public void run() {
                System.out.println("buzz");
            }
        };
        Runnable printFizzBuzz = new Runnable() {
            public void run() {
                System.out.println("fizzbuzz");
            }
        };
        IntConsumer printNumber = new IntConsumer() {
            public void accept(int x) {
                System.out.println(x);
            }
        };

        FizzBuzz sol = new FizzBuzz(5);

        // Threads for each operation
        Thread fizzThread = new Thread(() -> {
            try {
                sol.fizz(printFizz);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });
        Thread buzzThread = new Thread(() -> {
            try {
                sol.buzz(printBuzz);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });
        Thread fizzBuzzThread = new Thread(() -> {
            try {
                sol.fizzbuzz(printFizzBuzz);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });
        Thread numberThread = new Thread(() -> {
            try {
                sol.number(printNumber);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });

        // Start the threads
        fizzThread.start();
        buzzThread.start();
        fizzBuzzThread.start();
        numberThread.start();
    }
}

File: Problems/Problem_1195_Fizz_Buzz_Multithreaded/python/FizzBuzz.py
from typing import Callable
from threading import Thread, Condition

class FizzBuzz:
    def __init__(self, n: int):
        self.n = n
        self.counter = 1
        self.condition = Condition()

    # printFizz() outputs "fizz"
    def fizz(self, printFizz: Callable[[], None]) -> None:
        while self.counter <= self.n:
            with self.condition:
                while self.counter <= self.n and (self.counter % 3 != 0 or self.counter % 5 == 0):
                    self.condition.wait()  # Esperar turno
                if self.counter > self.n:
                    break
                printFizz()
                self.counter += 1
                self.condition.notify_all()  # Notificar a otros hilos

    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz: Callable[[], None]) -> None:
        while self.counter <= self.n:
            with self.condition:
                while self.counter <= self.n and (self.counter % 3 == 0 or self.counter % 5 != 0):
                    self.condition.wait()  # Esperar turno
                if self.counter > self.n:
                    break
                printBuzz()
                self.counter += 1
                self.condition.notify_all()  # Notificar a otros hilos

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz: Callable[[], None]) -> None:
        while self.counter <= self.n:
            with self.condition:
                while self.counter <= self.n and (self.counter % 3 != 0 or self.counter % 5 != 0):
                    self.condition.wait()  # Esperar turno
                if self.counter > self.n:
                    break
                printFizzBuzz()
                self.counter += 1
                self.condition.notify_all()  # Notificar a otros hilos

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber: Callable[[int], None]) -> None:
        while self.counter <= self.n:
            with self.condition:
                while self.counter <= self.n and (self.counter % 3 == 0 or self.counter % 5 == 0):
                    self.condition.wait()  # Esperar turno
                if self.counter > self.n:
                    break
                printNumber(self.counter)
                self.counter += 1
                self.condition.notify_all()  # Notificar a otros hilos


if __name__ == "__main__":
    
    # Funciones para imprimir
    printFizz = lambda: print("fizz")
    printBuzz = lambda: print("buzz")
    printFizzBuzz = lambda: print("fizzbuzz")
    printNumber = lambda x: print(x)

    # Crear instancia de FizzBuzz
    sol = FizzBuzz(15)

    # Crear hilos
    threadA = Thread(target=sol.fizz, args=(printFizz,))
    threadB = Thread(target=sol.buzz, args=(printBuzz,))
    threadC = Thread(target=sol.fizzbuzz, args=(printFizzBuzz,))
    threadD = Thread(target=sol.number, args=(printNumber,))

    # Iniciar hilos
    threadA.start()
    threadB.start()
    threadC.start()
    threadD.start()

    # Esperar a que todos los hilos terminen
    threadA.join()
    threadB.join()
    threadC.join()
    threadD.join()




